app/controllers/katello/api/v2/alternate_content_sources_controller.rb

```diff
@@ -117,12 +119,33 @@ def refresh
     protected
 
     def acs_params
-      keys = [:name, :label, :description, {smart_proxy_ids: []}, {smart_proxy_names: []}, :content_type, :alternate_content_source_type, :use_http_proxies]
-      keys += [:base_url, {subpaths: []}, :upstream_username, :upstream_password, :ssl_ca_cert_id, :ssl_client_cert_id, :ssl_client_key_id, :verify_ssl] if params[:action] == 'create' || @alternate_content_source&.custom? || @alternate_content_source&.rhui?
-      keys += [{product_ids: []}] if params[:action] == 'create' || @alternate_content_source&.simplified?
+      # In the past, this method flitered acs keys depending on what type the
+      # acs repo was. This caused silent failures. We're now passing everything
+      # to the model layer to mitigate.
+      keys = [
+        :name, :label, :description, {smart_proxy_ids: []}, {smart_proxy_names: []}, :content_type,
+        :alternate_content_source_type, :use_http_proxies, :base_url, {subpaths: []}, :upstream_username,
+        :upstream_password, :ssl_ca_cert_id, :ssl_client_cert_id, :ssl_client_key_id, :verify_ssl, {product_ids: []}
+      ]
+
       params.require(:alternate_content_source).permit(*keys).to_h.with_indifferent_access
     end
 
+    def check_params_for_invalid_updates
+      # Check parameters which cannot be validated at the model level, throwing
+      # errors where neccessary
+
+      # Check that the combination of params[:product_ids] and ACS type is allowed:
+      #            | simplified | custom  | rhui
+      # -----------+------------+---------+---------
+      # nil        | ok         | ok      | ok
+      # []         | ok         | invalid | invalid
+      # [foo, ...] | ok         | invalid | invalid
+      unless @alternate_content_source&.simplified? || params[:product_ids].nil?
+        (fail HttpErrors::UnprocessableEntity, "Products must remain blank for ACS of type #{@alternate_content_source&.alternate_content_source_type}")
```

Comment:
I'm not sure how Rails deals with associations in cases like this. Can this be removed as well and just rely on the existing validation in the model `validates :products, if: -> { custom? || rhui? }, absence: true` ? Or does that not work?

If we have to rely on the controller, it's not ideal because that would prevent changes in the API, but you could still do something like 
```
my_acs.product_ids = [42, 43]
```
in Rails console and it wouldn't complain.

Reply:
@qcjames53 This is the only unresolved thread here. It still seems like this could be covered with a custom validation in the model. were you able to try this?